File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/pathfinders/controllers/MZControllerMoveFlying.java
/*
package com.offz.spigot.mobzy.pathfinders.controllers;

import com.offz.spigot.mobzy.mobs.types.FlyingMob;
import net.minecraft.server.v1_15_R1.AxisAlignedBB;
import net.minecraft.server.v1_15_R1.ControllerMove;
import net.minecraft.server.v1_15_R1.MathHelper;

public class MZControllerMoveFlying extends ControllerMove {
    private final FlyingMob mob;
    private int j;

    public MZControllerMoveFlying(FlyingMob flyingMob) {
        super(flyingMob);
        this.mob = flyingMob;
    }

    public void a() {
        if (this.h == Operation.MOVE_TO) {
            double dx = this.b - this.mob.locX;
            double dy = this.c - this.mob.locY;
            double dz = this.d - this.mob.locZ;
            double distance = dx * dx + dy * dy + dz * dz;
            if (this.j-- <= 0) {
                this.j += this.mob.getRandom().nextInt(5) + 2;
                distance = (double) MathHelper.sqrt(distance);
//                if (this.hasLineOfSight(this.b, this.c, this.d, distance)) {
                    Double speed = mob.getStaticBuilder().getMovementSpeed();
                    if (speed == null)
                        speed = 0.1;
                    mob.motX += dx / distance * speed;
                    mob.motY += dy / distance * speed;
                    mob.motZ += dz / distance * speed;
//                } else {
//                    this.h = Operation.WAIT;
//                }
            }
        }
    }

    private boolean hasLineOfSight(double dx, double dy, double dz, double distance) {
        double x = (dx - this.mob.locX) / distance;
        double y = (dy - this.mob.locY) / distance;
        double z = (dz - this.mob.locZ) / distance;
        AxisAlignedBB axisalignedbb = this.mob.getBoundingBox();

        for (int i = 1; (double) i < distance; ++i) {
            axisalignedbb = axisalignedbb.d(x, y, z);
            if (!this.mob.world.getCubes(this.mob, axisalignedbb)) {
                return false;
            }
        }

        return true;
    }
}*/

File: src/main/java/com/mineinabyss/mobzy/pathfinders/controllers/MZControllerMoveFlying.kt
package com.mineinabyss.mobzy.pathfinders.controllers

import com.mineinabyss.mobzy.mobs.types.FlyingMob
import net.minecraft.server.v1_15_R1.AxisAlignedBB
import net.minecraft.server.v1_15_R1.ControllerMove
import net.minecraft.server.v1_15_R1.MathHelper
import net.minecraft.server.v1_15_R1.Vec3D
import kotlin.random.Random

/**
 * From EntityGhast's controller
 * TODO stop relying on NMS code
 */
class MZControllerMoveFlying(private val mob: FlyingMob) : ControllerMove(mob) {
    private var j = 0
    override fun a() {
        if (h == Operation.MOVE_TO && j-- <= 0) {
            j += Random.nextInt(2, 7)
            var vec3d = Vec3D(b - mob.locX, c - mob.locY, d - mob.locZ)
            vec3d = vec3d.d()
            if (this.hasLineOfSight(vec3d, MathHelper.f(vec3d.f()))) {
                val speed: Double = mob.staticTemplate.movementSpeed ?: 0.1
                mob.mot = mob.mot.e(vec3d.a(speed))
            } else h = Operation.WAIT
        }
    }

    fun stopNavigation() {
        h = Operation.WAIT
    }

    private fun hasLineOfSight(vec3d: Vec3D, distance: Int): Boolean {
        var axisalignedbb: AxisAlignedBB = mob.boundingBox
        for (j in 1 until distance) {
            axisalignedbb = axisalignedbb.b(vec3d)
            if (!mob.world.getCubes(mob, axisalignedbb)) {
                return false
            }
        }
        return true
    }


}
